concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,241 @@
+    /// first allocation.
+    heap_start:  UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to
+    /// `heap_start` if there are no more active allocations.
+    allocations: UnsafeCell<usize>,
+    /// Stores the last address given out, and allows for a small improvement
+    /// over the always-increasing bump allocator. Namely that if an item
+    /// `a` is allocated and deallocated without any allocations in between,
+    /// then the memory of `a` is reused for the next allocation.
+    last_alloc:  UnsafeCell<usize>,
+}
+
+impl BumpAllocator {
+    /// Create a new [`BumpAllocator`].
+    ///
+    /// This must be a `const` method for it to be used as a global allocator,
+    /// where it is called staticly. Since we cannot know the start and end
+    /// locations of the heap on compile time, several values are initialized to
+    /// a dummy value `0` and updated appropriately during the first allocation.
+    ///
+    /// # Safety
+    /// - Can only be used in single-threaded environments. The [`Sync`]
+    ///   implementation is only invalid and is only there to support setting
+    ///   this as the global allocator.
+    pub const unsafe fn new() -> Self {
+        Self {
+            // Initialized to the dummy value `0`.
+            next:        UnsafeCell::new(0),
+            // Initialized to the dummy value `0`.
+            heap_start:  UnsafeCell::new(0),
+            // Initialized to the dummy value `0`, which is checked during first initialization.
+            heap_end:    UnsafeCell::new(0),
+            allocations: UnsafeCell::new(0),
+            // Initialized to the dummy value `0`.
+            // Must be set to the same initial address as `next`.
+            last_alloc:  UnsafeCell::new(0),
+        }
+    }
+
+    /// Grow the memory by `delta` pages, each of size [PAGE_SIZE].
+    ///
+    /// If successful, it returns the previous number of memory pages.
+    /// Otherwise, it returns [`ERROR_PAGE_COUNT`] to indicate out of memory.
+    fn grow_memory(&self, delta: PageCount) -> PageCount {
+        // The first argument refers to the index of memory to return the size of.
+        // Currently, Wasm only supports a single slot of memory, so `0` must always be
+        // used. Source: https://doc.rust-lang.org/beta/core/arch/wasm32/fn.memory_grow.html
+        PageCount(core::arch::wasm32::memory_grow(0, delta.0))
+    }
+
+    /// Get the size of memory in terms of pages.
+    ///
+    /// The memory has three sections in the following order:
+    ///
+    ///  - The data section.
+    ///  - The stack.
+    ///  - The heap.
+    ///
+    /// To get the start location of the heap, use `__heap_base`.
+    fn size(&self) -> PageCount {
```

Comment:
Would it be possible to rename this to `memory_size` or `linear_memory_size` since I got confused for a bit thinking this is the size of the heap.

Reply:
Yep. Renamed to `memory_size` to match the actual name in wasm. I also renamed `grow_memory` to `memory_grow` for the same reason.